docs/ADR-61-merkle-proofed-entities.md

```diff
@@ -0,0 +1,238 @@
+needed to verify the entity against a Merkle Tree.
+
+The `merkleProof` property will be an object with the following schema:
+
+```json
+{
+  "type": "object",
+  "properties": {
+    "proof": {
+      "type": "array",
+      "items": {
+        "type": "string"
+      }
+    },
+    "index": {
+      "type": "number"
+    },
+    "hashingKeys": {
+      "type": "array",
+      "items": {
+        "type": "string"
+      }
+    },
+    "treeHashingMethod": {
+      "type": "string"
+    },
+    "entityHashingMethod": {
+      "type": "string"
+    },
+    "entityHash": {
+      "type": "string"
+    }
+  }
+}
+```
+
+Where `proof` is the array of hashes of nodes needed for the verification, `index` is the index of the entity in the Merkle Tree, `hashingKeys` is the array of keys of the entity metadata that
+will be used when computing the hash of the entity, and `treeHashingMethod` is the hashing method used to build the Merkle Tree.
+
+The `hashingKeys` property exists with the sole purpose of making this solution more flexible, allowing the entity metadata to change while being strict on the required properties when validating the hash.
+The property works by **whitelisting the keys at root level** that will be included in the object to be hashed.
+
+> :warning: **Keys order**: The hashingKeys will determine the order of the keys on the object to be hashed. Changing the order will result in a different hash.
+
+The `entityHashingMethod` will determine the algorithm the entity will be hashed with and the `entityHash` is a pre-computed hash of the entity that will be used as a previous check against the
+entity hashing procedure before verifying the entity against the Merkle Tree.
+
+Example of a Wearable entity metadata:
+
+```typescript
+{
+  id: "urn:decentraland:off-chain:base-avatars:aviatorstyle",
+  name: "Aviator Style",
+  description: "aDescription",
+  image: "image.png",
+  thumbnail: "thumbnail.png",
+  collectionAddress: "0x06012c8cf97bead5deae237070f9587f8e7a266d",
+  rarity: "unique",
+  data: {
+    tags: ["male", "man", "base-wearable"],
+    category: "eyewear",
+    replaces: [],
+    hides: [],
+    representations: [
+      {
+        bodyShapes: ["urn:decentraland:off-chain:base-avatars:BaseMale"],
+        mainFile: "M_Eyewear_AviatorStyle.glb",
+        overrideReplaces: [],
+        overrideHides: [],
+        contents: ["M_Eyewear_AviatorStyle.glb"],
+      },
+    ],
+  },
+  i18n: [{ code: "en", text: "Aviator Style" }],
+  createdAt: 1646935739,
+  updatedAt: 1646935739,
+  metrics: {
+    triangles: 0,
+    materials: 0,
+    meshes: 0,
+    bodies: 0,
+    entities: 0,
+    textures: 0,
+  },
+}
+```
+
+Example of a Third Party wearable entity metadata with a merkleProof:
+
+```typescript
+{
+  id: "urn:decentraland:off-chain:base-avatars:aviatorstyle",
+  name: "Aviator Style",
+  description: "aDescription",
+  image: "image.png",
+  thumbnail: "thumbnail.png",
```

Comment:
I haven't included the redundant content information, that is:
```typescript
content: {
  'thumbnail.png': 'someHash'
}
```
Should we include it here or should we assume that it will always be hashed?

Reply:
I think that is implementation dependent, "content" will simply be a key after all. Its content can vary from entity to entity. I think the proposal already covers the basics to ensure a good security level.

But yes, for wearables I'd definitely add the `"content": { key: hash,... }`